### PR TITLE
Added FButton.icon for a single icon button

### LINKED ISSUE #121: Add `FButton.icon` constructor
It'll be nice if there was a constructor for creating a icon button. The current button constructor that accepts a prefix and suffix icon ends up introducing additional padding.

See https://ui.shadcn.com/docs/components/button#icon

### PR DESCRIPTION
**Pull request by sommye-ctr**
**Describe the changes**
This adds a `FButton.icon` constructor which will be used to create buttons having a single icon as their content.
Resolves #121 

**Checklist**
- [x]  I have read the [CONTRIBUTING.md](../CONTRIBUTING.md).
- [x]  I have included the relevant unit/golden tests.
- [x] I have included the relevant samples.
- [x] I have updated the documentation accordingly.

### CODE REVIEW

File: docs/pages/docs/button.mdx [hunk 1]
```diff
@@ -103,7 +103,7 @@ FButton.raw(
   </Tabs.Tab>
 </Tabs>
 
-### With Icon
+### With Text+Icon
```
Comment (Pante) [inline docs/pages/docs/button.mdx:106]:
```suggestion
### With Text and Icon
```

File: forui/lib/src/widgets/button/button.dart [hunk 2]
```diff
@@ -99,6 +99,18 @@ class FButton extends StatelessWidget {
           label: label,
         );
 
+  /// Creates a [FButton] that contains only an [icon]
```
Comment (Pante) [inline forui/lib/src/widgets/button/button.dart:102]:
```suggestion
  /// Creates a [FButton] that contains only an [icon].
```

File: forui/lib/src/widgets/divider.dart [hunk 3]
```diff
@@ -168,6 +168,11 @@ final class FDividerStyle with Diagnosticable {
       ..add(DiagnosticsProperty('padding', padding))
       ..add(ColorProperty('color', color))
       ..add(DoubleProperty('width', width));
+
+    FButton.icon(
+      onPress: () {},
+      icon: FButtonIcon(icon: FAssets.icons.arrowRight),
+    );
```
Comment (Pante) [inline forui/lib/src/widgets/divider.dart:175]:
```suggestion
```

I think this snippet got added by accident.

File: forui/lib/src/widgets/button/button.dart [hunk 4]
```diff
@@ -99,6 +99,18 @@ class FButton extends StatelessWidget {
...
+  FButton.icon({
+    required this.onPress,
+    required Widget icon,
```
Comment (Pante) [inline forui/lib/src/widgets/button/button.dart:105]: Should we rename this to `child` @kawaijoe ? The [usage](https://github.com/forus-labs/forui/pull/138/files#diff-36884b437c4b1ac665bf24853e5fffe851ad98cd8753c981f6690c7fb241e798R146-R152) seems a little weird.

File: forui/lib/src/widgets/button/button_content.dart [hunk 5]
```diff
@@ -37,6 +37,22 @@ final class _FButtonContent extends StatelessWidget {
...
 }
 
+final class _FButtonIconContent extends StatelessWidget {
+  final Widget icon;
+
+  const _FButtonIconContent({required this.icon});
+
+  @override
+  Widget build(BuildContext context) {
+    final (:style, enabled: _) = FButton._of(context);
```
Comment (Pante) [inline forui/lib/src/widgets/button/button_content.dart:47]: I think we should introduce a new `FButtonIconStyle`. In general, each content should have its own style.

**Review by Pante — CHANGES_REQUESTED** [review]
Generally looks decent! One small issue is that I think the icon content should have its own style.

File: forui/lib/src/widgets/button/button.dart (hunk 4 above)
Comment (kawaijoe) [reply forui/lib/src/widgets/button/button.dart:105]: I think `child` makes sense here

File: forui/lib/src/widgets/button/button_content.dart (hunk 5 above)
Comment (sommye-ctr) [reply forui/lib/src/widgets/button/button_content.dart:47]: The properties be all same as `FButtonCustomStyle`?

Comment (Pante) [reply forui/lib/src/widgets/button/button_content.dart:47]: I think it's more similar to `FButtonContentStyle`.

Comment (sommye-ctr) [reply forui/lib/src/widgets/button/button_content.dart:47]: What should be in the inherit method parameters? The icon content style would be only containing padding (and not text styles as in `FButtonContentStyle`), so should the padding be directly given in the inherit method like

`padding = const EdgeInsets.symmetric(
          horizontal: 16,
          vertical: 12.5,
        )`

or should it be retrieved from any theme property?

Comment (sommye-ctr) [reply forui/lib/src/widgets/button/button_content.dart:47]: Yes that is correct, I'm asking in the `inherit` method in the `FButtonIconContentStyle`, what should be the parameters? If the padding be hard-coded as in `FButtonContentStyle` then it wouldn't require any method parameters.

Comment (Pante) [reply forui/lib/src/widgets/button/button_content.dart:47]: I think there isn't a need for a `FButtonIconContentStyle.inherit(...)` constructor. A normal constructor that accepts a padding, with a default value, should suffice.

Comment (sommye-ctr) [reply forui/lib/src/widgets/button/button_content.dart]: Is it fine now?

**Comment by Bartozo** [comment]
Just tried the library today and my first thought was that it is missing basic components such as  IconButton  😅 Glad to see that someone has already started working on this. Any chances to push this new component soon?

File: forui/lib/src/widgets/button/button.dart [hunk 6]
```diff
@@ -262,6 +274,82 @@ class FButtonCustomStyle extends FButtonStyle with Diagnosticable {
...
+          runtimeType == other.runtimeType &&
+          enabledBoxDecoration == other.enabledBoxDecoration &&
+          disabledBoxDecoration == other.disabledBoxDecoration &&
+          content == other.content &&
+          icon == other.icon;
+
+  @override
+  int get hashCode => enabledBoxDecoration.hashCode ^ disabledBoxDecoration.hashCode ^ content.hashCode ^ icon.hashCode;
+}
+
```
Comment (Pante) [inline forui/lib/src/widgets/button/button.dart:352]: I think there's a slight misunderstanding, `FButtonIconContentStyle` should be a field in `FButtonCustomStyle`, i.e. `iconContent`.

**Review by Pante — CHANGES_REQUESTED** [review]

File: forui/lib/src/widgets/button/button.dart (hunk 6 above)
Comment (sommye-ctr) [reply forui/lib/src/widgets/button/button.dart:352]: So there is a `FButtonIconContentStyle` which just contains a padding property for the icon-content. The `FButtonIconContentStyle` is also a property in `FButtonCustomStyle` named as `iconContent` which can be given by the user. This is then retrieved in `_FButtonIconContent` from the `FButton._of(...)` method and used. Is this right?

Comment (Pante) [reply forui/lib/src/widgets/button/button.dart:352]: > So there is a `FButtonIconContentStyle` which just contains a padding property for the icon-content. The `FButtonIconContentStyle` is also a property in `FButtonCustomStyle` named as `iconContent` which can be given by the user. This is then retrieved in `_FButtonIconContent` from the `FButton._of(...)` method and used. Is this right?

Yup sounds about right!

**Comment by sommye-ctr** [comment]
Did most of the changes as mentioned

**Comment by Pante** [comment]
I think you accidentally undid the changes in your merge commit. 😅

**Comment by sommye-ctr** [comment]
> I think you accidentally undid the changes in your merge commit. 😅

Ugh! I must have set to auto merge in IDE. WTH is going on

**Comment by sommye-ctr** [comment]
It should be fine now ig

File: forui/lib/src/widgets/button/button.dart [hunk 7]
```diff
@@ -230,12 +246,99 @@ class FButtonCustomStyle extends FButtonStyle with Diagnosticable {
...
+  @override
+  void debugFillProperties(DiagnosticPropertiesBuilder properties) {
+    super.debugFillProperties(properties);
+    properties
+      ..add(DiagnosticsProperty('enabledBoxDecoration', enabledBoxDecoration))
+      ..add(DiagnosticsProperty('disabledBoxDecoration', disabledBoxDecoration))
+      ..add(DiagnosticsProperty('content', content))
+      ..add(DiagnosticsProperty('icon', icon))
+      ..add(
+        DiagnosticsProperty<FButtonIconContentStyle>('iconContent', iconContent),
```
Comment (Pante) [inline forui/lib/src/widgets/button/button.dart:268]:
```suggestion
        DiagnosticsProperty('iconContent', iconContent),
```

File: forui/lib/src/widgets/button/button.dart [hunk 8]
```diff
@@ -200,12 +212,16 @@ class FButtonCustomStyle extends FButtonStyle with Diagnosticable {
   /// The icon's style.
   final FButtonIconStyle icon;
 
+  /// The icon-content's style
```
Comment (Pante) [inline forui/lib/src/widgets/button/button.dart:215]:
```suggestion
  /// The icon content's style.
```

File: forui/lib/src/widgets/button/button.dart [hunk 9]
```diff
@@ -230,12 +246,99 @@ class FButtonCustomStyle extends FButtonStyle with Diagnosticable {
...
+  int get hashCode =>
+      enabledBoxDecoration.hashCode ^
+      disabledBoxDecoration.hashCode ^
+      content.hashCode ^
+      icon.hashCode ^
+      iconContent.hashCode;
+}
+
+/// [FButton] style for icon-button.
+class FButtonIconStyle extends FButtonStyle with Diagnosticable {
```
Comment (Pante) [inline forui/lib/src/widgets/button/button.dart:293]: The `FButtonIconStyle` is no longer needed.

File: forui/lib/src/widgets/button/button_content.dart [hunk 10]
```diff
@@ -37,6 +37,59 @@ final class _FButtonContent extends StatelessWidget {
...
+
+    return Padding(
+      padding: style.iconContent.padding,
+      child: child,
+    );
+  }
+}
+
+/// [FButton] icon's style.
```
Comment (Pante) [inline forui/lib/src/widgets/button/button_content.dart:56]:
```suggestion
/// [FButton] icon content's style.
```

File: forui/lib/src/widgets/button/button_content.dart [hunk 11]
```diff
@@ -37,6 +37,59 @@ final class _FButtonContent extends StatelessWidget {
...
+        padding: padding ?? this.padding,
+      );
+
+  @override
+  bool operator ==(Object other) =>
+      identical(this, other) ||
+      other is FButtonIconContentStyle && runtimeType == other.runtimeType && padding == other.padding;
+
+  @override
+  int get hashCode => padding.hashCode;
```
Comment (Pante) [inline forui/lib/src/widgets/button/button_content.dart:84]: Please move this below the `debugFillProperties(...)` method.

File: forui/lib/src/widgets/button/button_content.dart [hunk 12]
```diff
@@ -37,6 +37,59 @@ final class _FButtonContent extends StatelessWidget {
...
+
+  @override
+  void debugFillProperties(DiagnosticPropertiesBuilder properties) {
+    super.debugFillProperties(properties);
+    properties.add(DiagnosticsProperty<EdgeInsets>('padding', padding));
```
Comment (Pante) [inline forui/lib/src/widgets/button/button_content.dart:89]:
```suggestion
    properties.add(DiagnosticsProperty('padding', padding));
```

File: forui/test/src/widgets/button/button_golden_test.dart [hunk 13]
```diff
@@ -136,7 +136,59 @@ void main() {
...
+              'button/$name-$variant-icon-disabled-button.png',
+            ),
+          );
+        });
       }
     }
   });
 }
+
+//flutter test --update-goldens /Users/somyemahajan/Projects/forui/forui/test/src/widgets/button/button_golden_test.dart
```
Comment (Pante) [inline forui/test/src/widgets/button/button_golden_test.dart:194]:
```suggestion
```

**Review by Pante — CHANGES_REQUESTED** [review]

**Comment by Pante** [comment]
Seems like there was a bug with our workflows which built code on the base branch instead of the head. I've updated the actions accordingly.

File: forui/lib/src/widgets/button/button_content.dart [hunk 14]
```diff
@@ -37,6 +37,59 @@ final class _FButtonContent extends StatelessWidget {
...
+  }
+}
+
+/// [FButton] icon content's style.
+class FButtonIconContentStyle with Diagnosticable {
+  /// The padding.
+  final EdgeInsets padding;
+
+  /// Creates a [FButtonIconContentStyle].
+  FButtonIconContentStyle({
```
Comment (Pante) [inline forui/lib/src/widgets/button/button_content.dart:62]: Can this constructor be constant? If so, we can provide `FCustomButtonStyle.iconContent` with a default value in its constructor.

Comment (sommye-ctr) [reply forui/lib/src/widgets/button/button_content.dart]: Done

**Comment by Pante** [comment]
LGTM! Thanks for the contribution.

**Review by Pante — APPROVED** [review]

**Pull request merged**